File: cpp/tests/unit-tests/SubscriptionQosTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "tests/utils/Gmock.h"
#include "tests/utils/Gtest.h"
#include <chrono>
#include <cstdint>

#include "joynr/OnChangeSubscriptionQos.h"
#include "joynr/OnChangeWithKeepAliveSubscriptionQos.h"
#include "joynr/PeriodicSubscriptionQos.h"

using namespace joynr;

class SubscriptionQosTest : public ::testing::Test
{
public:
    SubscriptionQosTest()
    {
    }

    std::int64_t nowPlusTimeSpan(const std::int64_t& timeSpan)
    {
        std::int64_t now = std::chrono::duration_cast<std::chrono::milliseconds>(
                                   std::chrono::system_clock::now().time_since_epoch())
                                   .count();
        return now + timeSpan;
    }
};

TEST_F(SubscriptionQosTest, createOnChangeSubscriptionQos)
{
    std::int64_t expiryDateMs = nowPlusTimeSpan(100000);
    std::int64_t publicationTtlMs = 2000;
    std::int64_t minIntervalMs = 100;
    OnChangeSubscriptionQos onChangeSubscriptionQos;
    onChangeSubscriptionQos.setExpiryDateMs(expiryDateMs);
    onChangeSubscriptionQos.setPublicationTtlMs(publicationTtlMs);
    onChangeSubscriptionQos.setMinIntervalMs(minIntervalMs);

    EXPECT_EQ(expiryDateMs, onChangeSubscriptionQos.getExpiryDateMs());
    EXPECT_EQ(publicationTtlMs, onChangeSubscriptionQos.getPublicationTtlMs());
    EXPECT_EQ(minIntervalMs, onChangeSubscriptionQos.getMinIntervalMs());
}

TEST_F(SubscriptionQosTest, createOnChangeSubscriptionQosWithValidity)
{
    std::int64_t validityMs = 100000;
    std::int64_t publicationTtlMs = 2000;
    std::int64_t minIntervalMs = 100;

    {
        std::int64_t expiryDateLowerBound = nowPlusTimeSpan(validityMs);
        OnChangeSubscriptionQos onChangeSubscriptionQos(
                validityMs, publicationTtlMs, minIntervalMs);
        std::int64_t expiryDateUpperBound = nowPlusTimeSpan(validityMs);
        onChangeSubscriptionQos.setPublicationTtlMs(publicationTtlMs);

        EXPECT_LE(expiryDateLowerBound, onChangeSubscriptionQos.getExpiryDateMs());
        EXPECT_GE(expiryDateUpperBound, onChangeSubscriptionQos.getExpiryDateMs());
        EXPECT_EQ(publicationTtlMs, onChangeSubscriptionQos.getPublicationTtlMs());
        EXPECT_EQ(minIntervalMs, onChangeSubscriptionQos.getMinIntervalMs());
    }
    {
        OnChangeSubscriptionQos onChangeSubscriptionQos;
        std::int64_t expiryDateLowerBound = nowPlusTimeSpan(validityMs);
        onChangeSubscriptionQos.setValidityMs(validityMs);
        std::int64_t expiryDateUpperBound = nowPlusTimeSpan(validityMs);
        onChangeSubscriptionQos.setPublicationTtlMs(publicationTtlMs);
        onChangeSubscriptionQos.setMinIntervalMs(minIntervalMs);

        EXPECT_LE(expiryDateLowerBound, onChangeSubscriptionQos.getExpiryDateMs());
        EXPECT_GE(expiryDateUpperBound, onChangeSubscriptionQos.getExpiryDateMs());
        EXPECT_EQ(publicationTtlMs, onChangeSubscriptionQos.getPublicationTtlMs());
        EXPECT_EQ(minIntervalMs, onChangeSubscriptionQos.getMinIntervalMs());
    }
}

TEST_F(SubscriptionQosTest, createPeriodicSubscriptionQos)
{
    std::int64_t expiryDateMs = nowPlusTimeSpan(100000);
    std::int64_t publicationTtlMs = 2000;
    std::int64_t periodMs = 800;
    std::int64_t alertAfterIntervalMs = 3000;

    PeriodicSubscriptionQos periodicSubscriptionQos;
    periodicSubscriptionQos.setExpiryDateMs(expiryDateMs);
    periodicSubscriptionQos.setPublicationTtlMs(publicationTtlMs);
    periodicSubscriptionQos.setPeriodMs(periodMs);
    periodicSubscriptionQos.setAlertAfterIntervalMs(alertAfterIntervalMs);

    EXPECT_EQ(expiryDateMs, periodicSubscriptionQos.getExpiryDateMs());
    EXPECT_EQ(publicationTtlMs, periodicSubscriptionQos.getPublicationTtlMs());
    EXPECT_EQ(periodMs, periodicSubscriptionQos.getPeriodMs());
    EXPECT_EQ(alertAfterIntervalMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
}

TEST_F(SubscriptionQosTest, createPeriodicSubscriptionQosWithValidity)
{
    std::int64_t validityMs = 1000;
    std::int64_t periodMs = 800;
    std::int64_t alertAfterIntervalMs = 3000;
    std::int64_t publicationTtlMs = 2000;
    {
        std::int64_t expiryDateLowerBound = nowPlusTimeSpan(validityMs);
        PeriodicSubscriptionQos periodicSubscriptionQos(
                validityMs, publicationTtlMs, periodMs, alertAfterIntervalMs);
        std::int64_t expiryDateUpperBound = nowPlusTimeSpan(validityMs);

        EXPECT_LE(expiryDateLowerBound, periodicSubscriptionQos.getExpiryDateMs());
        EXPECT_GE(expiryDateUpperBound, periodicSubscriptionQos.getExpiryDateMs());
        EXPECT_EQ(periodMs, periodicSubscriptionQos.getPeriodMs());
        EXPECT_EQ(alertAfterIntervalMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
    }
    {
        PeriodicSubscriptionQos periodicSubscriptionQos;
        std::int64_t expiryDateLowerBound = nowPlusTimeSpan(validityMs);
        periodicSubscriptionQos.setValidityMs(validityMs);
        std::int64_t expiryDateUpperBound = nowPlusTimeSpan(validityMs);
        periodicSubscriptionQos.setPeriodMs(periodMs);
        periodicSubscriptionQos.setAlertAfterIntervalMs(alertAfterIntervalMs);

        EXPECT_LE(expiryDateLowerBound, periodicSubscriptionQos.getExpiryDateMs());
        EXPECT_GE(expiryDateUpperBound, periodicSubscriptionQos.getExpiryDateMs());
        EXPECT_EQ(periodMs, periodicSubscriptionQos.getPeriodMs());
        EXPECT_EQ(alertAfterIntervalMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
    }
}

TEST_F(SubscriptionQosTest, createOnChangeWithKeepAliveSubscriptionQos)
{
    std::int64_t expiryDateMs = nowPlusTimeSpan(100000);
    std::int64_t publicationTtlMs = 2000;
    std::int64_t alertAfterIntervalMs = 4000;
    std::int64_t maxIntervalMs = 3000;
    std::int64_t minIntervalMs = 100;

    OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos;
    onChangeWithKeepAliveSubscriptionQos.setExpiryDateMs(expiryDateMs);
    onChangeWithKeepAliveSubscriptionQos.setPublicationTtlMs(publicationTtlMs);
    onChangeWithKeepAliveSubscriptionQos.setMaxIntervalMs(maxIntervalMs);
    onChangeWithKeepAliveSubscriptionQos.setMinIntervalMs(minIntervalMs);
    onChangeWithKeepAliveSubscriptionQos.setAlertAfterIntervalMs(alertAfterIntervalMs);

    EXPECT_EQ(expiryDateMs, onChangeWithKeepAliveSubscriptionQos.getExpiryDateMs());
    EXPECT_EQ(publicationTtlMs, onChangeWithKeepAliveSubscriptionQos.getPublicationTtlMs());
    EXPECT_EQ(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
    EXPECT_EQ(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMinIntervalMs());
    EXPECT_EQ(alertAfterIntervalMs, onChangeWithKeepAliveSubscriptionQos.getAlertAfterIntervalMs());
}

TEST_F(SubscriptionQosTest, alertAfterIntervalAdjustedIfSmallerThanPeriod)
{
    std::int64_t periodMs = 5000;
    std::int64_t alertAfterIntervalMs = 1000;

    PeriodicSubscriptionQos periodicSubscriptionQos;
    periodicSubscriptionQos.setPeriodMs(periodMs);
    periodicSubscriptionQos.setAlertAfterIntervalMs(alertAfterIntervalMs);

    EXPECT_EQ(periodMs, periodicSubscriptionQos.getPeriodMs());
    EXPECT_EQ(periodMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
}

TEST_F(SubscriptionQosTest, alertAfterIntervalAdjustedIfSmallerThanMaxInterval)
{
    std::int64_t alertAfterIntervalMs = 4000;
    std::int64_t maxIntervalMs = 5000;

    OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos;
    onChangeWithKeepAliveSubscriptionQos.setMaxIntervalMs(maxIntervalMs);
    onChangeWithKeepAliveSubscriptionQos.setAlertAfterIntervalMs(alertAfterIntervalMs);

    EXPECT_EQ(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
    EXPECT_EQ(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getAlertAfterIntervalMs());
}

TEST_F(SubscriptionQosTest, maxIntervalAdjustedIfSmallerThanMinInterval)
{
    std::int64_t maxIntervalMs = 1000;
    std::int64_t minIntervalMs = 2000;

    OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos;
    onChangeWithKeepAliveSubscriptionQos.setMaxIntervalMs(maxIntervalMs);
    onChangeWithKeepAliveSubscriptionQos.setMinIntervalMs(minIntervalMs);

    EXPECT_EQ(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
    EXPECT_EQ(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMinIntervalMs());
}

File: java/javaapi/src/test/java/joynr/SubscriptionQosTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package joynr;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import org.junit.Before;
import org.junit.Test;

import io.joynr.pubsub.SubscriptionQos;

public class SubscriptionQosTest {

    @Before
    public void setUp() {
    }

    @Test
    public void createOnChangeSubscriptionQos() throws Exception {
        long expiryDateMs = System.currentTimeMillis() + 100000;
        long publicationTtlMs = 2000;
        long minIntervalMs = 100;
        OnChangeSubscriptionQos onChangeSubscriptionQos = new OnChangeSubscriptionQos().setExpiryDateMs(expiryDateMs)
                                                                                       .setPublicationTtlMs(publicationTtlMs)
                                                                                       .setMinIntervalMs(minIntervalMs);
        assertEquals(expiryDateMs, onChangeSubscriptionQos.getExpiryDateMs());
        assertEquals(publicationTtlMs, onChangeSubscriptionQos.getPublicationTtlMs());
        assertEquals(minIntervalMs, onChangeSubscriptionQos.getMinIntervalMs());
    }

    @Test
    public void createOnChangeSubscriptionQosWithValidity() throws Exception {
        long validityMs = 100000;

        long lowerBound = System.currentTimeMillis() + validityMs;
        SubscriptionQos onChangeSubscriptionQos = new OnChangeSubscriptionQos().setValidityMs(validityMs);
        long upperBound = System.currentTimeMillis() + validityMs;

        assertTrue(onChangeSubscriptionQos.getExpiryDateMs() >= lowerBound);
        assertTrue(onChangeSubscriptionQos.getExpiryDateMs() <= upperBound);
    }

    @Test
    public void createPeriodicSubscriptionQos() throws Exception {
        long expiryDateMs = System.currentTimeMillis() + 100000;
        long publicationTtlMs = 2000;
        long periodMs = 800;
        long alertAfterIntervalMs = 3000;
        PeriodicSubscriptionQos periodicSubscriptionQos = new PeriodicSubscriptionQos().setExpiryDateMs(expiryDateMs)
                                                                                       .setPublicationTtlMs(publicationTtlMs)
                                                                                       .setPeriodMs(periodMs)
                                                                                       .setAlertAfterIntervalMs(alertAfterIntervalMs);
        assertEquals(expiryDateMs, periodicSubscriptionQos.getExpiryDateMs());
        assertEquals(publicationTtlMs, periodicSubscriptionQos.getPublicationTtlMs());
        assertEquals(periodMs, periodicSubscriptionQos.getPeriodMs());
        assertEquals(alertAfterIntervalMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
    }

    @Test
    public void createPeriodicSubscriptionQosWithValidity() throws Exception {
        long validityMs = 1000;

        long lowerBound = System.currentTimeMillis() + validityMs;
        SubscriptionQos periodicSubscriptionQos = new PeriodicSubscriptionQos().setValidityMs(validityMs);
        long upperBound = System.currentTimeMillis() + validityMs;

        assertTrue(periodicSubscriptionQos.getExpiryDateMs() >= lowerBound);
        assertTrue(periodicSubscriptionQos.getExpiryDateMs() <= upperBound);
    }

    @Test
    public void createOnChangeWithKeepAliveSubscriptionQos() throws Exception {
        long expiryDateMs = System.currentTimeMillis() + 100000;
        long publicationTtlMs = 2000;
        long alertAfterIntervalMs = 4000;
        long maxIntervalMs = 3000;
        long minIntervalMs = 100;
        OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos = new OnChangeWithKeepAliveSubscriptionQos().setExpiryDateMs(expiryDateMs)
                                                                                                                              .setPublicationTtlMs(publicationTtlMs)
                                                                                                                              .setMaxIntervalMs(maxIntervalMs)
                                                                                                                              .setMinIntervalMs(minIntervalMs)
                                                                                                                              .setAlertAfterIntervalMs(alertAfterIntervalMs);
        assertEquals(expiryDateMs, onChangeWithKeepAliveSubscriptionQos.getExpiryDateMs());
        assertEquals(publicationTtlMs, onChangeWithKeepAliveSubscriptionQos.getPublicationTtlMs());
        assertEquals(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
        assertEquals(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMinIntervalMs());
        assertEquals(alertAfterIntervalMs, onChangeWithKeepAliveSubscriptionQos.getAlertAfterIntervalMs());

    }

    @Test
    public void alertAfterIntervalAdjustedIfSmallerThanPeriod() throws Exception {
        long periodMs = 5000;
        long alertAfterIntervalMs = 1000;
        PeriodicSubscriptionQos periodicSubscriptionQos = new PeriodicSubscriptionQos().setPeriodMs(periodMs)
                                                                                       .setAlertAfterIntervalMs(alertAfterIntervalMs);
        assertEquals(periodMs, periodicSubscriptionQos.getPeriodMs());
        assertEquals(periodMs, periodicSubscriptionQos.getAlertAfterIntervalMs());
    }

    @Test
    public void alertAfterIntervalAdjustedIfSmallerThanMaxInterval() throws Exception {
        long alertAfterIntervalMs = 4000;
        long maxIntervalMs = 5000;
        OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos = new OnChangeWithKeepAliveSubscriptionQos().setMaxIntervalMs(maxIntervalMs)
                                                                                                                              .setAlertAfterIntervalMs(alertAfterIntervalMs);
        assertEquals(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
        assertEquals(maxIntervalMs, onChangeWithKeepAliveSubscriptionQos.getAlertAfterIntervalMs());
    }

    @Test
    public void maxIntervalAdjustedIfSmallerThanMinInterval() throws Exception {
        long minIntervalMs = 2000;
        long maxIntervalMs = 1000;
        OnChangeWithKeepAliveSubscriptionQos onChangeWithKeepAliveSubscriptionQos = new OnChangeWithKeepAliveSubscriptionQos().setMaxIntervalMs(maxIntervalMs)
                                                                                                                              .setMinIntervalMs(minIntervalMs);
        assertEquals(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMinIntervalMs());
        assertEquals(minIntervalMs, onChangeWithKeepAliveSubscriptionQos.getMaxIntervalMs());
    }
}
